### PR TITLE
RE-1421 Configurable Timeout

### PR DESCRIPTION
**Pull request by hughsaunders**
Add timeout to configuration, so deployers can choose their own request timeout.

### CODE REVIEW

**Comment by hughsaunders** [comment]
I haven't quite finished testing this, but please review and I'll merge tomorrow if testing is successful and no objections.

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java [hunk 1]
```diff
@@ -231,6 +235,16 @@ public NodePool(String connectionString,
         return acceptedNodes;
     }
 
+    public String getRequestTimeout() {
+        return requestTimeout;
+    }
+
+    public final void setRequestTimeout(String requestTimeout) {
+        // ensure value can be parsed as an integer
+        Integer.parseInt(requestTimeout);
```
Comment (dealako) [inline src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: May want to ensure value is a positive value > 0.  After parsing, `if Integer.parseInt(requestTimeout) < 1 throw new IllegalArgumentException("Timeout value must be positive.");` A zero or negative value is undesirable.  Maybe add javadoc to indicate the value is in seconds??  I see we have form validation below to ensure that it's positive, but we can also add checks on the implementation.  We can easily add a quick unit test to the method for this too...

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java [hunk 2]
```diff
@@ -145,6 +143,11 @@ public static CuratorFramework createZKConnection(String connectionString,
      */
     private String requestor;
 
+    /**
+     * Timeout for a node request
+     */
+    private String requestTimeout;
```
Comment (dealako) [inline src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: Store request timeout as String or Integer?  I guess it's a matter of style - convert the value earlier or later?

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java [hunk 3]
```diff
@@ -439,14 +458,21 @@ void provisionNode(Label label, Task task, int timeoutInSec) throws Exception {
...
+    /**
+     * Cleanup a failed NodeRequest and associated Queue item.
+     *
+     * @param request
+     * @param task
+     */
+    void cleanup(NodeRequest request, Task task) {
+        requests.remove(request);
+        Jenkins.getInstance().getQueue().cancel(task);
+    }
```
Comment (dealako) [inline src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: +1 for consolidating cleanup :-)

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java [hunk 4]
```diff
@@ -499,6 +525,20 @@ public FormValidation doCheckLabelPrefix(@QueryParameter String labelPrefix) {
             }
         }
 
+        public FormValidation doCheckRequestTimeout(@QueryParameter String requestTimeout) {
+            LOG.log(Level.INFO, "doCheckRequestTimeout: {0}", requestTimeout);
```
Comment (dealako) [inline src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: Change to `Level.FINE` once done testing (?).

**Review by dealako — COMMENTED** [review]
Like the changes.  Just a few comments to consider (or discard) :-).

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java [hunk 5]
```diff
@@ -499,6 +525,20 @@ public FormValidation doCheckLabelPrefix(@QueryParameter String labelPrefix) {
...
+            String error = "Request Timeout must be a whole number greater than zero.";
```
Comment (bdelliott) [inline src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: One could make this allow an empty value and then just apply a default timeout to make this field optional.

File: src/test/java/com/rackspace/jenkins_nodepool/TestTEst.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import static org.junit.Assert.fail;
+import org.junit.Test;
+
+/**
+ *
+ * @author hughsaunders
+ */
+public class TestTEst {
```
Comment (bdelliott) [inline src/test/java/com/rackspace/jenkins_nodepool/TestTEst.java]: Um?

**Review by bdelliott — APPROVED** [review]
Seems reasonable

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java (hunk 2 above)
Comment (hughsaunders) [reply src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: Done

Comment (hughsaunders) [reply src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: Done

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java (hunk 3 above)
Comment (hughsaunders) [reply src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: 😄

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java (hunk 4 above)
Comment (hughsaunders) [reply src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: Removed log statements

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java (hunk 5 above)
Comment (hughsaunders) [reply src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: I think providing a default achieves the same thing while being more explicit about the preset value.

File: src/test/java/com/rackspace/jenkins_nodepool/TestTEst.java (hunk 6 above)
Comment (hughsaunders) [reply src/test/java/com/rackspace/jenkins_nodepool/TestTEst.java]: OOOPS

File: src/main/java/com/rackspace/jenkins_nodepool/NodePool.java (hunk 1 above)
Comment (hughsaunders) [reply src/main/java/com/rackspace/jenkins_nodepool/NodePool.java]: Done

**Pull request merged**